picard/file.py

```diff
@@ -296,14 +299,39 @@ def _save_and_rename(self, old_filename, metadata):
             return None
         new_filename = old_filename
         if not config.setting["dont_write_tags"]:
-            save = partial(self._save, old_filename, metadata)
+            old_filename_temp = old_filename+".temp"
+
+            # Make a copy to modify
+            shutil.copy(old_filename, old_filename_temp)
```

Comment:
So the idea is to work on a copy, which is safer for sure.
But it has drawbacks: it copies the whole file before, and if metadata to be written requires a rewrite, that will be 2 times, it can be awfully slow on network file systems. Alone it deserves an option for the user to choose between safety & performance.

Also `shutil.copy()` will not preserve owner/group/permissions.
Temporary file has to be created with a dedicated method (we can't assume `somefile.temp` doesn't exist).

This code has to be tested on all platforms and various situations: code creating/moving/copying is very likely to have different behaviors in different envs.

Reply:
Yes, I'm not too happy with this for the performance reasons alone. This for sure can make a huge difference. E.g. in many situations Picard only has to rewrite the tag section. If this is the case it safes a lot of IO, especially for large files such as FLAC. If we make creating copies mandatory this involves a lot more data writing.

Reply:
Isn't `shutil.copy2()` that doesn't carry permissions? Other metadata (dates, etc) can be copied with `shutil.copystat`.
I don't see why we can't assume .temp doesn't exist. A temporary file is a temporary file.

I would never trade data integrity for performance, but I see the point. Better keep the copy in memory, perform the metadata writes in memory, check if works in memory and then dump to the disk as the .temp, and use shutil.move to replace the original. I'm going to work on that.

Reply:
> I don't see why we can't assume .temp doesn't exist. A temporary file is a temporary file.

Don't assume, ensure it doesn't exist.
Here you just name a file with suffix `.temp` based on old file name, and such file might exist already, this is why all temporary files creation methods are ensuring no such file exists (and try to return a file descriptor to avoid race conditions), and use names to reduce the risk of conflict.



> Isn't `shutil.copy2()` that doesn't carry permissions? Other metadata (dates, etc) can be copied with `shutil.copystat`.

Nope, `shutil.copy2()` will actually copy more "metadata", but none of those are copying user & group information for example (that's the normal behavior on *nix for copies).

So if we want to preserve those you have to copy them over from original file, and hope the program can actually set them (it may not).

https://docs.python.org/3/library/shutil.html#shutil.copy

Reply:
Changed the logic to keep the contents of the file in memory while performing the saving. If the saving fails, rewrite the original contents, and in case it succeeds the memory is freed.

It's pretty hard to further reduce IO/syscalls due to the interface between Picard and mutagen (path instead of buffer/file handler/etc).

Reply:
Actually you can pass a file descriptor to mutagen, that works just fine. I even experimented with this in https://github.com/phw/picard/commit/28b36a779a14da8df3a04511c006d21eff13d7ff . The code is a bit hackish at places and I never found the time to finish it, but it works.

Reply:
> Changed the logic to keep the contents of the file in memory while performing the saving. If the saving fails, rewrite the original contents, and in case it succeeds the memory is freed.

The only really method is to write new file to a temporary file, then do an "atomic" rename. If you store in memory, and the program crashes you'll not be able to rewrite originals.

So to implement this properly:

- create an uniquely named temporary file in the same directory of the target file, keep it open, use its file descriptor
- copy permissions and times from original file before writing any data
- write data to temp file, ensure it is ok (no error + optionally, checksum)
- rename temp file to file "atomically" (if the system allows it, POSIX systems do)

If the program or system or machine crashes during write, a temp file will be left over.
If write ops fail, old file will not be overwritten.
If permissions cannot be set properly, it will error.

Reply:
Absolutely correct on that. Still working on that, but I had to slow down a bit due to work and just wanted to commit the checking thing before I forgot about it.